src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -178,14 +181,84 @@ class SetupPyChrootRequest:
     py2: bool  # Whether to use py2 or py3 package semantics.
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class SetupKwargs:
+    """The keyword arguments to the `setup()` function in the generated `setup.py`."""
+
+    json_str: str
+
+    def __init__(
+        self, kwargs: Mapping[str, Any], *, address: Address, _allow_banned_keys: bool = False
+    ) -> None:
+        super().__init__()
+        if "version" not in kwargs:
+            raise ValueError(f"Missing a `version` kwarg in the `provides` field for {address}.")
+
+        if not _allow_banned_keys:
+            for arg in {"data_files", "package_dir", "package_data", "packages"}:
+                if arg in kwargs:
+                    raise ValueError(
+                        f"{arg} cannot be set in the `provides` field for {address}, but it was "
+                        f"set to {kwargs[arg]}. Pants will dynamically set the value for you."
+                    )
+
+        # We convert to a `str` so that this type can be hashable. We don't use `FrozenDict`
+        # because it would require that all values are immutable, and we may have lists and
+        # dictionaries as values. It's too difficult/clunky to convert those all, then to convert
+        # them back out of `FrozenDict`.
+        self.json_str = json.dumps(kwargs, sort_keys=True)
+
+    @memoized_property
+    def kwargs(self) -> Dict[str, Any]:
+        return cast(Dict[str, Any], json.loads(self.json_str))
+
+    @property
+    def name(self) -> str:
+        return cast(str, self.kwargs["name"])
+
+    @property
+    def version(self) -> str:
+        return cast(str, self.kwargs["version"])
+
+
+@union
+@dataclass(frozen=True)  # type: ignore[misc]
+class SetupKwargsPluginRequest(ABC):
+    """An entry point to customize the kwargs used for the `setup()` function, rather than the
```

Comment:
This isn't what is usually meant by an "entry point". And again the use of "plugin" is confusing. Maybe a more helpful description would be "A request for kwargs to pass to the setup() function. By default, Pants will pass the kwargs provided in the BUILD file unchanged. To customize this behavior, subclass `SetupKwargsRequest`, register the rule... "  etc etc.

Reply:
See above. This is only relevant for plugins. This type is not used if there are no plugins.

Reply:
Maybe this is the source of what seems to me like clunkiness. Could the default behavior just be uniform with the custom behavior, both being union members?

Reply:
And either way, this is not an "entry point"...

Reply:
> Could the default behavior just be uniform with the custom behavior, both being union members?

That's precisely how I first implemented it, until changing it based on your suggestion this morning. This new implementation is simpler than before.

Even if we made the default case a union member, we would still need custom logic to distinguish between the default case and plugins. By definition, the default case should be "True" for `is_valid()` for every target type. This means we could end up with the default working as the same time as a plugin implementation. We can't have that because we don't know which of the 2 to use.

> this is not an "entry point"...

Genuine question: why not? It is a "hook" into our `setup-py` goal.